pkgs/development/libraries/qt-6/modules/qtwebengine.nix

```diff
@@ -227,6 +216,45 @@ qtModule rec {
 
   requiredSystemFeatures = [ "big-parallel" ];
 
+  # NOTE buildPhase ignores NIX_BUILD_CORES
+  # and uses all available cpu cores
+  #
+  # limiting cores by
+  #   export NINJAFLAGS="-j32 -l32"
+  # causes the build error
+  #   internal compiler error: Segmentation fault
+  #
+  # https://bugreports.qt.io/browse/QTBUG-103573
+  #
+  # honor NIX_BUILD_CORES in recursive ninja calls
+  # https://bugreports.qt.io/browse/QTBUG-95176
+  #
+  # based on ninjaBuildPhase in
+  # pkgs/development/tools/build-managers/ninja/setup-hook.sh
+  #
+  # this must run before cmake
+  # to set NINJAFLAGS for qtwebengine/cmake/Functions.cmake
+  #
+  /*
+  preConfigure = ''
+    local buildCores=1
+
+    # Parallel building is enabled by default.
+    if [ "''${enableParallelBuilding-1}" ]; then
+        buildCores="$NIX_BUILD_CORES"
+    fi
+
+    local flagsArray=(
+        -j$buildCores -l$NIX_BUILD_CORES
+        $ninjaFlags "''${ninjaFlagsArray[@]}"
+    )
+
+    # honor NIX_BUILD_CORES in recursive ninja calls
+    export NINJAFLAGS="''${flagsArray[@]}"
+    echo "preConfigure: setting NINJAFLAGS: $NINJAFLAGS"
+  '';
+  */
```

Comment:
im working on this in my branch [qt6-631-debug-qtwebengine](https://github.com/milahu/nixpkgs/tree/qt6-631-debug-qtwebengine)

the main build process is ninja
so currently im exploring [ninja-tokenpool](https://github.com/ninja-build/ninja/issues/1139#issuecomment-461750356)
to limit the job count across multiple ninja processes

todo: add jobclients to python and node builders in qtwebengine

* src/3rdparty/chromium/third_party/blink/renderer/bindings/scripts/bind_gen/task_queue.py
* src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup-plugin-terser/rollup-plugin-terser.js

by default, these builders use all cpu cores

Reply:
> in rare cases, the workers dont release their tokens back to the jobserver

What is the chance that the number of tokens drops to zero or a very low number? It would be annoying if that could hang a build.

Reply:
here, the problem are bugs in my jobclients
the load slowly drops to 1 = no parallelism (but the build does not hang)

but its also a well-known weakness/bug in the jobserver protocol
https://github.com/milahu/gnumake-tokenpool/issues/1#issuecomment-1170227666

Reply:
I guess that almost any time a job fails the whole build is going to fail anyways so it shouldn't be too big of a problem. In that I case I think it would be fine with a warning. But if you think you can fix it that is even better.

Reply:
fixed in https://github.com/milahu/gnumake-tokenpool/commit/247c42b6ac517c3e4b45de0af22d57ad45ec44a7

now the max load is 32 == NIX_BUILD_CORES

![qt6-qtwebengine-631-build log 2022-07-03 14-08-13 gz load dat plt](https://user-images.githubusercontent.com/12958815/177112524-1fd1b6c2-e8fb-45e9-aeef-e15ff0ef3f4e.png)

Reply:
@kevincox 

> I tried building e206f8755a457b84cdc44dd43b0d6d3518181334 and it has stalled out as well

same here : (

with [qt6-631-debug-qtwebengine-2](https://github.com/milahu/nixpkgs/tree/qt6-631-debug-qtwebengine-2) i see that the javascript workers are blocking ...

```
 load  Load  rss  time spr cpr proc
 20.0  18.1 9869   471 123   2 bash 1: bash -e /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh # {'cwd': '/build/qtwebengine-everywhere-src-6.3.1/build', 'exe': '/nix/store/j9rc786ylq8cid4zgcn6idknilkbd4ax-bash-5.1-p16/bin/bash'}
 20.0  18.1 9863   425 121   1   ninja 13856: ninja -j32 -l32 --tokenpool-master # {'exe': '/nix/store/k38101zidsncnv0qvcia82n1nmvkb57a-ninja-tokenpool-unstable-2022-02-13/bin/ninja'}
 20.0  18.1 9858   378 120  40     ninja 17319: ninja -C /build/qtwebengine-everywhere-src-6.3.1/build/src/core/Release/x86_64 QtWebEngineCore # {'cwd': '/build/qtwebengine-everywhere-src-6.3.1/build/src/core/Release/x86_64', 'exe': '/nix/store/k38101zidsncnv0qvcia82n1nmvkb57a-ninja-tokenpool-unstable-2022-02-13/bin/ninja'}
  1.0   1.1  956   229   2   1       python3 31547: python3 ../../../../../src/3rdparty/chromium/third_party/blink/renderer/bindings/scripts/generate_bindings.py --web_idl_database gen/third_party/blink/renderer/bindings/web_idl_database.pickle --root_src_dir ../../../../../src/3rdparty/chromium/ --root_gen_dir gen --output_reldir core=third_party/blink/renderer/bindings/core/v8/ --output_reldir modules=third_party/blink/renderer/bindings/modules/v8/ --format_generated_files enumeration callback_function callback_interface dictionary interface namespace typedef union # {'exe': '/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/bin/python3.10'}
  1.0   1.0  790   228   1   0         python3 31861: python3 ../../../../../src/3rdparty/chromium/third_party/blink/renderer/bindings/scripts/generate_bindings.py --web_idl_database gen/third_party/blink/renderer/bindings/web_idl_database.pickle --root_src_dir ../../../../../src/3rdparty/chromium/ --root_gen_dir gen --output_reldir core=third_party/blink/renderer/bindings/core/v8/ --output_reldir modules=third_party/blink/renderer/bindings/modules/v8/ --format_generated_files enumeration callback_function callback_interface dictionary interface namespace typedef union # {'exe': '/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/bin/python3.10'}
  0.0   0.0  132   225   3   1       python3 32929: python3 ../../../../../src/3rdparty/chromium/third_party/node/node.py ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --silent --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/entrypoints/formatter_worker/FormatterActions.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/entrypoints/formatter_worker/FormatterActions.js --configDCHECK # {'exe': '/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/bin/python3.10'}
  0.0   0.0  122   224   2   1         node 32930: node ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/entrypoints/formatter_worker/FormatterActions.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/entrypoints/formatter_worker/FormatterActions.js --configDCHECK # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
  0.0   0.0   25   224   1   0           node 32995: node /build/qtwebengine-everywhere-src-6.3.1/src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/jest-worker/build/workers/processChild.js # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
  0.0   0.0  149   223   3   1       python3 33114: python3 ../../../../../src/3rdparty/chromium/third_party/node/node.py ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --silent --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/core/platform/platform.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/core/platform/platform.js --configDCHECK # {'exe': '/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/bin/python3.10'}
  0.0   0.0  139   223   2   1         node 33136: node ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/core/platform/platform.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/core/platform/platform.js --configDCHECK # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
  0.0   0.0   25   223   1   0           node 33512: node /build/qtwebengine-everywhere-src-6.3.1/src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/jest-worker/build/workers/processChild.js # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
  0.0   0.0  136   223   3   1       python3 33246: python3 ../../../../../src/3rdparty/chromium/third_party/node/node.py ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --silent --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/models/heap_snapshot_model/heap_snapshot_model.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/models/heap_snapshot_model/heap_snapshot_model.js --configDCHECK # {'exe': '/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/bin/python3.10'}
  0.0   0.0  126   223   2   1         node 33267: node ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/rollup/dist/bin/rollup --config ../../../../../src/3rdparty/chromium/third_party/devtools-frontend/src/scripts/build/rollup.config.js --input gen/third_party/devtools-frontend/src/front_end/models/heap_snapshot_model/heap_snapshot_model.prebundle.js --file gen/third_party/devtools-frontend/src/front_end/models/heap_snapshot_model/heap_snapshot_model.js --configDCHECK # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
  0.0   0.0   25   222   1   0           node 33645: node /build/qtwebengine-everywhere-src-6.3.1/src/3rdparty/chromium/third_party/devtools-frontend/src/node_modules/jest-worker/build/workers/processChild.js # {'exe': '/nix/store/2ph85rmc3j8n5k679y4zp0qvqzrmwwkn-nodejs-16.15.0/bin/node'}
```

note the time column. the workers hang since 200 seconds with average load zero (2nd column)

blocked by ... https://github.com/ninja-build/ninja/issues/2164